### DISCUSSION TITLE
C 语言调用 Zig (FFI 完整用例)

### DISCUSSION THREAD

**Discussion by hhstore** [zigcc/forum#12] [General]
## 如题: 

> 背景: 

- https://ziglang.org/learn/overview/#export-functions-variables-and-types-for-c-code-to-depend-on
- ✅ 官方用例, 因为 zig v0.6.0 之后, 编译选项废弃 `-femit-h`, 导致不可用. 
- ✅ 如下示例, 完整解决. 并更新 `build.zig` 编译配置, 适应当前最新版本 `zig v0.10`

> 完整用例:

https://github.com/better-zig/learn-zig/blob/main/packages/zig-to-c/readme.md

> 说明:

- ✅ 从 c 语言 调用 zig 方法, FFI 完整示例. (这样 zig FFI 给其他语言调用, 就走通了)
- ✅ zig v0.6 之后, 去掉了 生成 .h 头文件的选项. 

> 步骤: 

1. 先基于 zig v0.6.0 docker 生成 .h 头文件.
2. 手动更改 build.zig.
3. 然后用 zig v0.10 正常编译. 

> 官方 issue:

- https://github.com/ziglang/zig/issues/6753
- 官方 issue 无人问津状态. 🤣

## 用例的使用场景: 

- 该例子是 C 语言调用 Zig. 
- 原理, 适用其他语言调用 Zig.
- 打通 zig FFI 作为 core lib 层, 平替 `rust` 的应用场景. 

## 关于 zig FFI 更多参考: 

- ✅ https://github.com/hhstore/blog/issues/384
- ✅ https://github.com/hhstore/blog/issues/362

**Comment by jiacai2050** [comment]
```
zig build-lib -fno-stage1 -femit-h zgo.zig
```
那个 issue 下有人回复了，高版本可以用这种方式生成 header 文件

**Reply by hhstore** [reply]
> ```
> zig build-lib -fno-stage1 -femit-h zgo.zig
> ```
> 
> 那个 issue 下有人回复了，高版本可以用这种方式生成 header 文件


看了下, 确实有回复了. 但是, 试了一下这个编译选项生成的 .h 头文件. 并不 work.

- 这个编译选项, 能生成 header 文件. 但包含大量冗余代码. 上面一个单文件, 生成1600+行. 
- 直接替换我上面的 .h header 文件编译, 会报错的. (可能要改一些编译选项)

> PS: 

- 按照官方并没有 closed 上述一组 issue, 我认为, 官方可能并没有真正解决. 
- 这个编译做法, 可能是不恰当使用. 
- v0.6 生成的头文件, 就只包含这个函数声明, 非常简单. 够用了. 
- 我把 `zig build-lib -fno-stage1 -femit-h` 生成结果也保留了. `mathtest2.h`, 可以对比二者差异.
